File: src/test/java/ev/projects/GameTest.java
package ev.projects;

import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.List;

import static ev.projects.PatternLineTest.patternLines;
import static org.junit.jupiter.api.Assertions.*;

public class GameTest {

    @Test
    void gameEndingExecutesOnFilledHorizontalLine() {
        Center center = new Center();
        center.addTile(Tile.BLUE);
        Player player = new PlayerMother().newPlayer(new WallMother().withAlmostCompletedHorizontalLine());
        Game game = new Game(List.of(player, new PlayerMother().newPlayer()), center, 0);
        game.clearFactoryDisplays();

        game.executeFactoryOfferPhaseWithCenter(Tile.BLUE, 0, 0);

        assertEquals(6, player.score());
    }

    @Test
    void wallTilingExecutesOnEmptyFactoryDisplaysAndCenter() {
        Center center = new Center();
        center.addTile(Tile.BLUE);
        PatternLine[] patternLines = patternLines();
        Player player = new Player(new Board(patternLines));
        Game game = new Game(List.of(player, new PlayerMother().newPlayer()), center, 0);
        game.clearFactoryDisplays();

        game.executeFactoryOfferPhaseWithCenter(Tile.BLUE, 0, 0);

        assertFalse(patternLines[0].isFilled());
    }

    @Test
    void playerTakesAndPlacesFactoryTiles() {
        PatternLine[] patternLines = patternLines();
        Floor floor = new Floor();
        Player player = new Player(new Board(patternLines, new Wall(), floor));
        Game game = new Game(List.of(player, new PlayerMother().newPlayer()), new Center(), 0);
        game.changeFactoryDisplay(0, Tile.RED, Tile.RED, Tile.RED, Tile.RED);
        GameController gameController = new GameController(game);

        gameController.takeTilesFromFactory(new FactoryTakingRequest(0, Tile.RED, 2, 1));

        assertEquals(-2, floor.score());
        assertEquals("R R", floor.toString());
        assertEquals(2, patternLines[1].tileCount());
        assertEquals(Tile.RED, patternLines[1].tile());
    }

    @Test
    void playerTakesAndPlacesCenterTiles() {
        PatternLine[] patternLines = patternLines();
        Floor floor = new Floor();
        Player player = new Player(new Board(patternLines, new Wall(), floor));
        Center center = new Center();
        Game game = new Game(List.of(player, new PlayerMother().newPlayer()), center, 1);
        game.changeFactoryDisplay(0, Tile.RED, Tile.RED, Tile.YELLOW, Tile.BLACK);
        game.executeFactoryOfferPhaseWithFactory(0, Tile.YELLOW, 0, 3);
        GameController gameController = new GameController(game);

        gameController.takeTilesFromCenter(new CenterTakingRequest(Tile.RED, 1, 0));

        assertEquals(-2, floor.score());
        assertEquals("M R", floor.toString());
        assertEquals(1, patternLines[0].tileCount());
        assertEquals(Tile.RED, patternLines[0].tile());
        assertEquals("K", center.toString());
    }

    @Test
    void playersTakeTilesInOrder() {
        Center center = new Center();
        PatternLine[] patternLines1 = patternLines();
        Player player1 = new Player(new Board(patternLines1));
        PatternLine[] patternLines2 = patternLines();
        Player player2 = new Player(new Board(patternLines2));
        Game game = new Game(List.of(player1, player2), center, 0);

        game.executeFactoryOfferPhaseWithFactory(0, game.factoryDisplays()[0].tiles()[0], 0, 0);
        assertTrue(patternLines1[0].tileCount() > 0);
        game.executeFactoryOfferPhaseWithFactory(1, game.factoryDisplays()[1].tiles()[0], 0, 0);
        assertTrue(patternLines2[0].tileCount() > 0);
        game.executeFactoryOfferPhaseWithFactory(2, game.factoryDisplays()[2].tiles()[0], 0, 1);
        assertTrue(patternLines1[1].tileCount() > 0);
        game.executeFactoryOfferPhaseWithFactory(3, game.factoryDisplays()[3].tiles()[0], 0, 1);
        assertTrue(patternLines2[1].tileCount() > 0);
    }

    @Test
    void gameIsDisplayed() {
        Lid lid = new Lid();
        Player player1 = new PlayerMother().newPlayer("Robert");
        Wall wall2 = new Wall();
        Floor floor2 = new Floor(lid);
        Player player2 = new PlayerMother().newPlayer(wall2, floor2, "Roger");
        Center center = new Center();
        Game game = new Game(List.of(player1, player2), center, 0, lid);
        game.changeFactoryDisplay(0, Tile.RED, Tile.RED, Tile.RED, Tile.BLUE);
        game.changeFactoryDisplay(1, Tile.RED, Tile.RED, Tile.YELLOW, Tile.BLUE);
        game.changeFactoryDisplay(2, Tile.YELLOW, Tile.YELLOW, Tile.YELLOW, Tile.YELLOW);
        game.changeFactoryDisplay(3, Tile.WHITE, Tile.WHITE, Tile.WHITE, Tile.WHITE);
        game.changeFactoryDisplay(4, Tile.BLACK, Tile.WHITE, Tile.WHITE, Tile.YELLOW);
        game.setBag(new Bag(initTilesInBag(new int[] {18, 14, 15, 19, 14})));
        game.executeFactoryOfferPhaseWithFactory(4, Tile.WHITE, 1, 2);
        floor2.add(Tile.RED, 1);
        floor2.add(Tile.YELLOW, 2);
        player2.takeTilesFromCenter(center, Tile.YELLOW, 0, 1);
        wall2.add(Tile.BLUE, 2);
        game.executeWallTilingPhase();

        System.out.println(new GameController(game).show());
        assertEquals("""
                        Factories: 1) B 3R 2) B Y 2R 3) 4Y 4) 4W 5) Empty
                        Center: K
                        Player Robert:
                        Score: 0
                        Has the starting player token.
                        Board:
                        Pattern lines:
                        E\s
                        E E\s
                        W E E\s
                        E E E E\s
                        E E E E E\s
                        Wall:
                        b y r k w\s
                        w b y r k\s
                        k w b y r\s
                        r k w b y\s
                        y r k w b\s
                        Floor: W
                        Player Roger:
                        Score: 0
                        Board:
                        Pattern lines:
                        E\s
                        Y E\s
                        E E E\s
                        E E E E\s
                        E E E E E\s
                        Wall:
                        b y r k w\s
                        w b y r k\s
                        k w B y r\s
                        r k w b y\s
                        y r k w b\s
                        Floor: R Y Y M
                        Bag: 18B 14Y 15R 19K 14W
                        Lid: Empty""",
                new GameController(game).show()
        );
    }

    private List<Tile> initTilesInBag(int[] amount) {
        List<Tile> tiles = new ArrayList<>();
        for (int i = 0; i < amount.length; i++) {
            for(int j = 0; j < amount[i]; j++) {
                tiles.add(Tile.values()[i]);
            }
        }
        return tiles;
    }
}

File: python/lib/test/game_test.py
import unittest

from azul.bag import Bag
from azul.board import Board
from azul.center import Center
from azul.floor import Floor
from azul.game import Game
from azul.lid import Lid
from azul.pattern_line import PatternLine
from azul.player import Player
from azul.tile import Tile
from azul.wall import Wall
from player_mother import PlayerMother
from wall_mother import WallMother


class TestGame(unittest.TestCase):
    def test_game_ending_executes_on_filled_horizontal_line(self):
        center = Center()
        center.add_tile(Tile(Tile.BLUE))
        player = PlayerMother().new_player(wall=WallMother.with_almost_completed_horizontal_line())
        game = Game([player, PlayerMother().new_player()], center, 0)
        game.clear_factory_displays()

        game.execute_factory_offer_phase_with_center(Tile(Tile.BLUE), 0, 0)

        self.assertEqual(6, player.score)

    def test_wall_tiling_executes_on_empty_factory_displays_and_center(self):
        center = Center()
        center.add_tile(Tile.BLUE)
        pattern_lines = self.pattern_lines()
        player = Player(Board(pattern_lines))
        game = Game([player, PlayerMother().new_player()], center, 0)
        game.clear_factory_displays()

        game.execute_factory_offer_phase_with_center(Tile(Tile.BLUE), 0, 0)

        self.assertFalse(pattern_lines[0].is_filled())

    def test_player_takes_and_places_factory_tiles(self):
        pattern_lines = self.pattern_lines()
        floor = Floor()
        player = Player(Board(pattern_lines, Wall(), floor))
        game = Game([player, PlayerMother().new_player()], Center(), 0)
        game.change_factory_display(0, [Tile.RED, Tile.RED, Tile.RED, Tile.RED])

        game.execute_factory_offer_phase_with_factory(0, Tile.RED, 2, 1)

        self.assertEqual(-2, floor.score())
        self.assertEqual("R R", str(floor))
        self.assertEqual(2, pattern_lines[1].tile_count)
        self.assertEqual(Tile.RED, pattern_lines[1].tile)

    def test_player_takes_and_places_center_tiles(self):
        pattern_lines = self.pattern_lines()
        floor = Floor()
        player = Player(Board(pattern_lines, Wall(), floor))
        center = Center()
        game = Game([player, PlayerMother().new_player()], center, 1)
        game.change_factory_display(0, [Tile.RED, Tile.RED, Tile.YELLOW, Tile.BLACK])
        game.execute_factory_offer_phase_with_factory(0, Tile.YELLOW, 1, 3)

        game.execute_factory_offer_phase_with_center(Tile.RED, 1, 0)

        self.assertEqual(-2, floor.score())
        self.assertEqual("M R", str(floor))
        self.assertEqual(1, pattern_lines[0].tile_count)
        self.assertEqual(Tile.RED, pattern_lines[0].tile)
        self.assertEqual("K", str(center))

    def test_players_take_tiles_in_order(self):
        center = Center()
        pattern_lines1 = self.pattern_lines()
        player1 = Player(Board(pattern_lines1))
        pattern_lines2 = self.pattern_lines()
        player2 = Player(Board(pattern_lines2))
        game = Game([player1, player2], center, 0)

        game.execute_factory_offer_phase_with_factory(0, game.factory_displays[0].tiles[0], 0, 0)
        self.assertTrue(pattern_lines1[0].tile_count > 0)

        game.execute_factory_offer_phase_with_factory(1, game.factory_displays[1].tiles[0], 0, 0)
        self.assertTrue(pattern_lines2[0].tile_count > 0)

        game.execute_factory_offer_phase_with_factory(2, game.factory_displays[2].tiles[0], 0, 1)
        self.assertTrue(pattern_lines1[1].tile_count > 0)

        game.execute_factory_offer_phase_with_factory(3, game.factory_displays[3].tiles[0], 0, 1)
        self.assertTrue(pattern_lines2[1].tile_count > 0)

    def test_game_is_displayed(self):
        lid = Lid()
        player1 = PlayerMother().new_player(wall=Wall(), floor=Floor(lid), name="Robert")
        wall2 = Wall()
        floor2 = Floor(lid)
        player2 = PlayerMother().new_player(wall=wall2, floor=floor2, name="Roger")
        center = Center()
        game = Game([player1, player2], center, 0, lid)
        game.change_factory_display(0, [Tile.RED, Tile.RED, Tile.RED, Tile.BLUE])
        game.change_factory_display(1, [Tile.RED, Tile.RED, Tile.YELLOW, Tile.BLUE])
        game.change_factory_display(2, [Tile.YELLOW, Tile.YELLOW, Tile.YELLOW, Tile.YELLOW])
        game.change_factory_display(3, [Tile.WHITE, Tile.WHITE, Tile.WHITE, Tile.WHITE])
        game.change_factory_display(4, [Tile.BLACK, Tile.WHITE, Tile.WHITE, Tile.YELLOW])
        game.set_bag(Bag(self.init_tiles_in_bag([18, 14, 15, 19, 14])))
        game.execute_factory_offer_phase_with_factory(4, Tile.WHITE, 1, 2)
        floor2.add(Tile.RED, 1)
        floor2.add(Tile.YELLOW, 2)
        player2.take_tiles_from_center(center, Tile.YELLOW, 0, 0)
        wall2.add(Tile.BLUE, 2)
        game.execute_wall_tiling_phase()

        json_object = game.json_object()

        self.assertEqual("{'W': 1, 'R': 1, 'Y': 2}", str(json_object["Lid"]))
        self.assertTrue("'Pattern lines': [[], [], ['W'], [], []]" in str(json_object["Players"]))
        self.assertTrue(
            "'Wall': [['b', 'Y', 'r', 'k', 'w'], ['w', 'b', 'y', 'r', 'k'], ['k', 'w', 'B', 'y', 'r'], ['r', 'k', 'w', 'b', 'y'], ['y', 'r', 'k', 'w', 'b']]"
                in str(json_object["Players"])
        )

    @staticmethod
    def init_tiles_in_bag(amounts):
        tiles = []
        for i, count in enumerate(amounts):
            tiles.extend([Tile(list(Tile)[i])] * count)
        return tiles

    @staticmethod
    def pattern_lines():
        return [PatternLine(1), PatternLine(2), PatternLine(3), PatternLine(4), PatternLine(5)]